ui/src/app/preferences/views/SSHKeyList/SSHKeyList.js

```diff
@@ -0,0 +1,189 @@
+import { Link } from "react-router-dom";
+import { useDispatch, useSelector } from "react-redux";
+import React, { useEffect, useState } from "react";
+
+import "./SSHKeyList.scss";
+import { sshkey as sshkeyActions } from "app/preferences/actions";
+import { sshkey as sshkeySelectors } from "app/preferences/selectors";
+import Button from "app/base/components/Button";
+import VanillaLink from "app/base/components/Link";
+import Loader from "app/base/components/Loader";
+import MainTable from "app/base/components/MainTable";
+import TableDeleteConfirm from "app/base/components/TableDeleteConfirm";
+
+const formatKey = key => {
+  const parts = key.split(" ");
+  if (parts.length === 3) {
+    return parts[2];
+  }
+  return `${key.slice(0, 20)}...`;
+};
+
+const groupBySource = sshkeys => {
+  const groups = new Map();
+  sshkeys.forEach(sshkey => {
+    const { keysource } = sshkey;
+    let source, id, groupId;
+    if (keysource) {
+      const { protocol, auth_id } = keysource;
+      groupId = `${protocol}/${auth_id}`;
+      id = auth_id;
+      source =
+        (protocol === "lp" && "Launchpad") ||
+        (protocol === "gh" && "GitHub") ||
+        protocol;
+    } else {
+      source = "Upload";
+      groupId = sshkey.id;
+    }
```

Comment:
Might be nice to have a few unittests for this logic, confirming that source is set correctly, or perhaps just another component test?